heartbeat/autodiscover/builder/hints/heartbeats.go

```
@@ -0,0 +1,211 @@
+	return builder.GetHintMapStr(hints, l.config.Key, montype)
+}
+
+func (l *heartBeat) getSchedule(hints common.MapStr) []string {
+	return builder.GetHintAsList(hints, l.config.Key, schedule)
+}
+
+func (l *heartBeat) getRawConfigs(hints common.MapStr) []common.MapStr {
+	return builder.GetHintAsConfigs(hints, l.config.Key)
+}
+
+func (l *heartBeat) getMonitors(hints common.MapStr) []common.MapStr {
+	raw := builder.GetHintMapStr(hints, l.config.Key, "")
+	if raw == nil {
+		return nil
+	}
+
+	var words, nums []string
+
+	for key := range raw {
+		if _, err := strconv.Atoi(key); err != nil {
+			words = append(words, key)
+			continue
+		} else {
+			nums = append(nums, key)
+		}
+	}
+
+	sort.Strings(nums)
+
+	var configs []common.MapStr
+	for _, key := range nums {
+		rawCfg, _ := raw[key]
+		if config, ok := rawCfg.(common.MapStr); ok {
+			configs = append(configs, config)
+		}
+	}
+
+	defaultMap := common.MapStr{}
+	for _, word := range words {
+		defaultMap[word] = raw[word]
+	}
+
+	if len(defaultMap) != 0 {
+		configs = append(configs, defaultMap)
+	}
+	return configs
+}
+
+func (l *heartBeat) getProcessors(hints common.MapStr) []common.MapStr {
+	return builder.GetConfigs(hints, "", "processors")
+}
+
+func (l *heartBeat) getHostsWithPort(hints common.MapStr, port int) ([]string, bool) {
+	var result []string
+	thosts := builder.GetHintAsList(hints, "", hosts)
+	// Only pick hosts that have ${data.port} or the port on current event. This will make
+	// sure that incorrect meta mapping doesn't happen
+	for _, h := range thosts {
+		if strings.Contains(h, "data.port") || strings.Contains(h, fmt.Sprintf(":%d", port)) ||
+			// Use the event that has no port config if there is a ${data.host}:9090 like input
+			(port == 0 && strings.Contains(h, "data.host")) {
+			result = append(result, h)
+		} else if port == 0 && !strings.Contains(h, ":") {
+			// For ICMP like use cases allow only host to be passed if there is no port
+			result = append(result, h)
+		}
```

What do you think about adding a default `else` case here letting the user know their config won't run? In general, it's good to avoid silent failures since they're hard to debug.

---

the reason why this is not defaulted to just `else` is because we dont want to create a `host` config unless we can make use of it meaningfully. in such scenarios we wont spin up a `monitor` because it fits neither host nor host-port patterns.

---

I think there was a bit of a misunderstanding here. I was saying that we should probably log an error in the `else` case.